src/api/base.ts

```diff
@@ -0,0 +1,146 @@
+import {
+  Status,
+  ModuleProfile,
+  Api,
+  PluginProfile,
+  ApiEventEmitter,
+  PluginApi,
+  PluginRequest,
+  ExtractKey,
+} from '../types'
+import { EventEmitter } from 'events'
+
+type Profile<T extends Api> = ModuleProfile<T> | PluginProfile<T>
+
+/**
+ * Create a Profile with default values
+ * @param profile The profile of the plugin
+ * @param mixinProfiles A list of profiles to mix in the base profile
+ */
+export function createProfile<T extends Api, U extends Api>(
+  profile: Profile<T>,
+  ...mixinProfiles: Partial<ModuleProfile<U>>[]
+): Profile<T> {
+  return {
+    ...profile,
+    methods: [
+      ...mixinProfiles.reduce((acc, mixin) => [ ...acc, ...(mixin.methods || []) ], []),
+      ...(profile.methods || [])
+    ],
+    events: [
+      ...mixinProfiles.reduce((acc, mixin) => [ ...acc, ...(mixin.events || []) ], []),
+      ...(profile.events || [])
+    ],
+    notifications: {
+      ...mixinProfiles.reduce((acc, mixin) => ({ ...acc, ...(mixin.notifications || {}) }), {} as any),
+      ...(profile.notifications || {}),
+    },
+  }
+}
+
+export const baseProfile = {
+  notifications: {
+    'theme': ['switchTheme']
+  }
+}
+
+export abstract class BaseApi<S, U extends Api> {
+  private initialState: S
+  protected state: S
+  protected requestQueue: Array<() => Promise<any>> = []
+  protected currentRequest: PluginRequest
+  public readonly profile: ModuleProfile<U> | PluginProfile<U>
+  public events: ApiEventEmitter<U>
+  public activate?(): Promise<void>
+  public deactivate?(): void
+  public render?(): HTMLElement
+
+  constructor(
+    profile: ModuleProfile<U> | PluginProfile<U>,
+    initialState?: S,
+  ) {
+    this.events = new EventEmitter() as ApiEventEmitter<U>
+    this.profile = createProfile(profile, baseProfile)
+    this.initialState = initialState || {} as S
+    this.state = { ...this.initialState }
+  }
+
+  //////////////////////
+  // STATE MANAGEMENT //
+  //////////////////////
+
+  /**
+   * Update one field of the state
+   * @param state The part of the state updated
+   */
+  public updateState(state: Partial<S>) {
+    this.state = { ...this.state, ...state }
+  }
+
+  /** Get the state or a part of it */
+  public getState(): S
+  public getState<K extends keyof S>(key: K): S[K]
+  public getState<R>(query: (store: S) => R): R
+  public getState<K extends keyof S, R>(
+    query?: ((state: S) => R) | K,
+  ): R | S | S[K] {
+    switch (typeof query) {
+      case 'function': return query(this.state)
+      case 'string': return this.state[query]
+      default: return this.state
+    }
+  }
+
+  /** Reset the state its initial value */
+  public resetState() {
+    this.state = this.initialState
+  }
+
+
+  /////////
+  // API //
+  /////////
+
+  /** Exports an Api interface */
+  public api(): PluginApi<U> {
```

Comment:
just a thought: should we change the name `PluginApi` at some point? because it seems to refer to a Plugin, although it is also valid for Module

Reply:
Yes we should. Any idea ? =)

Reply:
`RuntimeAPI` maybe (but perhaps not in this PR)